src/sfptpd_clock.c

```diff
@@ -2484,4 +2571,163 @@ bool sfptpd_clock_is_active(const struct sfptpd_clock *clock)
+			calibration->error = true;
+		}
+	}
+
+	/* Check effect of adjusting freq for each clock */
+	for (i = 0; i < count; i++) {
+		struct sfptpd_clock *clock = snapshot[i];
+		struct sfptpd_clock_calibration *calibration = calibrations + i;
+		struct sfptpd_timespec mod_offset;
+
+		if (calibration->error)
+			continue;
+
+		/* Apply temporary offset */
+		sfptpd_time_add(&mod_offset, &calibration->offset, &test_adjustment);
+		rc = sfptpd_clock_adjust_time(clock, &test_adjustment);
+		if (rc != 0) {
+			ERROR("clock %s: dedup2, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+		usleep(settle);
+
+		/* Check new offset */
+		rc = sfptpd_clock_compare(clock, sys, &t);
+		if (rc != 0) {
+			ERROR("clock %s: dedup3, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+		if (!sfptpd_time_equal_within(&t, &mod_offset, &threshold)) {
+			INFO("clock %s: deduplication: did not take test adjustment\n",
+			     clock->short_name);
+			calibration->inert = true;
+		}
+
+		/* Check if offset impacts other clocks */
+		for (j = 0; j < count; j++) {
+			struct sfptpd_clock *other = snapshot[j];
+			if (i == j)
+				continue;
+			if (calibrations[j].error)
+				continue;
+
+			rc = sfptpd_clock_compare(other, sys, &t);
+			if (rc != 0) {
+				ERROR("clock %s: dedup4, %s\n",
+				      other->short_name, strerror(rc));
+				calibrations[j].error = true;
+			}
+			if (!sfptpd_time_equal_within(&t, &calibrations[j].offset, &threshold)) {
+				struct sfptpd_timespec mod2;
+				TRACE_L4("clock %s: clock %s: has changed\n",
+				     clock->short_name, other->short_name);
+				sfptpd_time_add(&mod2, &calibrations[j].offset, &test_adjustment);
+				if (sfptpd_time_equal_within(&t, &mod2, &threshold)) {
+					INFO("clock %s: deduplication: %s governs %s\n",
+					     clock->short_name, clock->short_name, other->short_name);
+					calibrations[j].governor = clock;
+				}
+			}
+		}
+
+		/* Restore original offset */
+		rc = sfptpd_clock_adjust_time(clock, &restore_adjustment);
+		if (rc != 0) {
+			ERROR("clock %s: dedup5, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+		usleep(settle);
+	}
+
+	NOTICE("clock: identification of duplicate phc clocks complete\n");
+
+	for (i = 0; i < count; i++) {
+		struct sfptpd_clock *clock = snapshot[i];
+		struct sfptpd_clock_calibration *calibration = calibrations + i;
+
+		sfptpd_interface_hw_timestamping_restore(&calibration->intf_query);
+
+		if (calibration->governor) {
+			sfptpd_interface_reassign_to_nic(clock->u.nic.device_idx,
+							 calibration->governor->u.nic.device_idx);
+			changes++;
+		} else if (calibration->inert) {
+			NOTICE("clock: marking inert clock %s as read only\n",
+			       clock->short_name);
+			clock->read_only = true;
+			changes++;
+		}
+	}
+
+	free(calibrations);
+finish:
+	free(snapshot);
+
+	if (changes) {
+		sfptpd_clock_rescan_interfaces();
+		sfptpd_interface_diagnostics(3);
```

Comment:
Why the magic number 3?

Reply:
That's logging level. So the symbolic version would be something like `TRACE_LEVEL_3`. Would that be an improvement?